json2html/jsonconv.py

```diff
@@ -31,42 +37,69 @@ def convert(self,**args):
 				self.json_input = json.dumps(self.json_input)
 		else:
 			raise Exception('Can\'t convert NULL!')
-		
-		
+
+
 		ordered_json = json.loads(self.json_input, object_pairs_hook=ordereddict.OrderedDict)
 		return self.htmlConvertor(ordered_json)
 
+	def columnHeadersFromListOfDicts(self,ordered_json):
+		if len(ordered_json) < 2:
+			return None
+		if not isinstance(ordered_json[0],dict):
+			return None
+		column_headers = ordered_json[0].keys()
+		for entry in ordered_json:
+			if not isinstance(entry,dict):
+				return None
+			if len(entry.keys()) != len(column_headers):
+				return None
+			for header in column_headers:
+				if not header in entry:
+					return None
+		return column_headers
 
+	#ordered_json needs to be a dict
 	def iterJson(self,ordered_json):
+		def markup(entry, parent_is_list=False):
+			if(isinstance(entry,unicode)):
+				return unicode(entry)
+			if(isinstance(entry,int) or isinstance(entry,float)):
+				return str(entry)
+			if(parent_is_list and isinstance(entry,list)==True):
+				#list of lists are not accepted
+				return ''
+			if(isinstance(entry,list)==True) and len(entry) == 0:
+				return ''
+			if(isinstance(entry,list)==True):
+				return '<ul><li>' + '</li><li>'.join([markup(child, parent_is_list=True) for child in entry]) + '</li></ul>'
+			if(isinstance(entry,dict)==True):
+				return self.iterJson(entry)
+			return '' #safety: don't do recursion over anything that we don't know about - iteritems() will most probably fail
+
 		global a
-		a=a+ "<table border=\"1\">" 
+		global table_attributes
+
+		table_init_markup = "<table %s>" %(table_attributes)
```

Comment:
ok, just committed another fix that readds the borders as default behavior. I’ve tested your json and to me it looks ok now. About my json: Here you go, it’s a doozy ;-).

{"errors": ["invalid frontend definition. Please check the domain and the characters in the name: frontend:myproject-oldtest4.myproject.com", "invalid frontend definition. Please check the domain and the characters in the name: frontend:myproject-oldtest5.myproject.com", "invalid frontend definition. Please check the domain and the characters in the name: frontend:myproject-oldtest6.myproject.com", "invalid frontend definition. Please check the domain and the characters in the name: frontend:ategra-test.myproject.com", "invalid frontend definition. Please check the domain and the characters in the name: frontend:myproject-oldtestjan1.myproject.com", "invalid frontend definition. Please check the domain and the characters in the name: frontend:www.myproject.com", "invalid frontend definition. Please check the domain and the characters in the name: frontend:acme.myproject.com", "invalid frontend definition. Please check the domain and the characters in the name: frontend:myproject.com", "invalid frontend definition. Please check the domain and the characters in the name: frontend:zh-bd.myproject.com", "invalid frontend definition. Please check the domain and the characters in the name: frontend:acme2.myproject.com", "invalid frontend definition. Please check the domain and the characters in the name: frontend:zh-vd.myproject.com", "invalid frontend definition. Please check the domain and the characters in the name: frontend:myproject-oldtest1.myproject.com", "invalid frontend definition. Please check the domain and the characters in the name: frontend:myproject-oldtest2.myproject.com", "invalid frontend definition. Please check the domain and the characters in the name: frontend:myproject-oldtest3.myproject.com", "ategra-test's metadata is invalid", "www's metadata is invalid", "acme's metadata is invalid", "myproject's metadata is invalid", "zh-bd's metadata is invalid", "acme2's metadata is invalid", "zh-vd's metadata is invalid", "no db backend defined for environment myproject-oldtestjan2. Client replication will not be available!", "no db backend defined for environment 09test1. Client replication will not be available!"], "environments": [{"status": "no-server-assigned", "web backends": [], "server": "undefined", "environment name": "myproject-oldtest4", "db backends": [], "email used for registration": "peter.muster@gmail.com"}, {"status": "no-server-assigned", "web backends": [], "server": "undefined", "environment name": "myproject-oldtest5", "db backends": [], "email used for registration": "peter.muster@gmail.com"}, {"status": "no-server-assigned", "web backends": [], "server": "undefined", "environment name": "myproject-oldtest7", "db backends": [], "email used for registration": "peter.muster@gmail.com"}, {"status": "no-server-assigned", "web backends": [], "server": "undefined", "environment name": "myproject-oldtest6", "db backends": [], "email used for registration": "peter.muster@gmail.com"}, {"status": "undefined", "web backends": [], "server": "undefined", "environment name": "ategra-test", "db backends": [], "email used for registration": "n/a"}, {"status": "no-server-assigned", "web backends": [], "server": "undefined", "environment name": "myproject-oldtestjan1", "db backends": [], "email used for registration": "peter.muster@gmail.com"}, {"status": "undefined", "web backends": [], "server": "undefined", "environment name": "www", "db backends": [], "email used for registration": "n/a"}, {"status": "undefined", "web backends": [], "server": "undefined", "environment name": "acme", "db backends": [], "email used for registration": "n/a"}, {"status": "no-server-assigned", "web backends": [], "server": "undefined", "environment name": "myproject-oldtest8", "db backends": [], "email used for registration": "peter.muster@gmail.com"}, {"status": "undefined", "web backends": [], "server": "undefined", "environment name": "myproject", "db backends": [], "email used for registration": "n/a"}, {"status": "undefined", "web backends": [], "server": "undefined", "environment name": "zh-bd", "db backends": [], "email used for registration": "n/a"}, {"status": "no-server-assigned", "web backends": [], "server": "undefined", "environment name": "myproject-oldtest10", "db backends": [], "email used for registration": "peter.muster@gmail.com"}, {"status": "undefined", "web backends": [], "server": "undefined", "environment name": "acme2", "db backends": [], "email used for registration": "n/a"}, {"status": "undefined", "web backends": [], "server": "undefined", "environment name": "zh-vd", "db backends": [], "email used for registration": "n/a"}, {"status": "no-server-assigned", "web backends": [], "server": "undefined", "environment name": "myproject-oldtest1", "db backends": [], "email used for registration": "peter.muster@gmail.com"}, {"status": "no-server-assigned", "web backends": [], "server": "undefined", "environment name": "myproject-oldtest2", "db backends": [], "email used for registration": "peter.muster@gmail.com"}, {"status": "no-server-assigned", "web backends": [], "server": "undefined", "environment name": "myproject-oldtest3", "db backends": [], "email used for registration": "peter.muster@gmail.com"}, {"status": "activated", "web backends": ["pgrid0.myproject-old.com:10001"], "server": "pgrid0", "environment name": "myproject-oldtestjan2", "db backends": [], "email used for registration": "peter.muster@gmail.com"}, {"status": "activated", "web backends": ["pgrid0.myproject-old.com:10002"], "server": "pgrid0", "environment name": "09test1", "db backends": [], "email used for registration": "peter.muster@gmail.com"}, {"status": "activated", "web backends": ["pgrid0.myproject-old.com:10003"], "server": "pgrid0", "environment name": "testnew20151", "db backends": ["pgrid0.myproject-old.com:11003"], "email used for registration": "peter.muster@gmail.com"}], "servers": ["pgrid0"]}
{"errors": ["invalid frontend definition. Please check the domain and the characters in the name: frontend:myproject-oldtest4.myproject.com", "invalid frontend definition. Please check the domain and the characters in the name: frontend:myproject-oldtest5.myproject.com", "invalid frontend definition. Please check the domain and the characters in the name: frontend:myproject-oldtest6.myproject.com", "invalid frontend definition. Please check the domain and the characters in the name: frontend:ategra-test.myproject.com", "invalid frontend definition. Please check the domain and the characters in the name: frontend:myproject-oldtestjan1.myproject.com", "invalid frontend definition. Please check the domain and the characters in the name: frontend:www.myproject.com", "invalid frontend definition. Please check the domain and the characters in the name: frontend:acme.myproject.com", "invalid frontend definition. Please check the domain and the characters in the name: frontend:myproject.com", "invalid frontend definition. Please check the domain and the characters in the name: frontend:zh-bd.myproject.com", "invalid frontend definition. Please check the domain and the characters in the name: frontend:acme2.myproject.com", "invalid frontend definition. Please check the domain and the characters in the name: frontend:zh-vd.myproject.com", "invalid frontend definition. Please check the domain and the characters in the name: frontend:myproject-oldtest1.myproject.com", "invalid frontend definition. Please check the domain and the characters in the name: frontend:myproject-oldtest2.myproject.com", "invalid frontend definition. Please check the domain and the characters in the name: frontend:myproject-oldtest3.myproject.com", "ategra-test's metadata is invalid", "www's metadata is invalid", "acme's metadata is invalid", "myproject's metadata is invalid", "zh-bd's metadata is invalid", "acme2's metadata is invalid", "zh-vd's metadata is invalid", "no db backend defined for environment myproject-oldtestjan2. Client replication will not be available!", "no db backend defined for environment 09test1. Client replication will not be available!"], "environments": [{"status": "no-server-assigned", "web backends": [], "server": "undefined", "environment name": "myproject-oldtest4", "db backends": [], "email used for registration": "peter.muster@gmail.com"}, {"status": "no-server-assigned", "web backends": [], "server": "undefined", "environment name": "myproject-oldtest5", "db backends": [], "email used for registration": "peter.muster@gmail.com"}, {"status": "no-server-assigned", "web backends": [], "server": "undefined", "environment name": "myproject-oldtest7", "db backends": [], "email used for registration": "peter.muster@gmail.com"}, {"status": "no-server-assigned", "web backends": [], "server": "undefined", "environment name": "myproject-oldtest6", "db backends": [], "email used for registration": "peter.muster@gmail.com"}, {"status": "undefined", "web backends": [], "server": "undefined", "environment name": "ategra-test", "db backends": [], "email used for registration": "n/a"}, {"status": "no-server-assigned", "web backends": [], "server": "undefined", "environment name": "myproject-oldtestjan1", "db backends": [], "email used for registration": "peter.muster@gmail.com"}, {"status": "undefined", "web backends": [], "server": "undefined", "environment name": "www", "db backends": [], "email used for registration": "n/a"}, {"status": "undefined", "web backends": [], "server": "undefined", "environment name": "acme", "db backends": [], "email used for registration": "n/a"}, {"status": "no-server-assigned", "web backends": [], "server": "undefined", "environment name": "myproject-oldtest8", "db backends": [], "email used for registration": "peter.muster@gmail.com"}, {"status": "undefined", "web backends": [], "server": "undefined", "environment name": "myproject", "db backends": [], "email used for registration": "n/a"}, {"status": "undefined", "web backends": [], "server": "undefined", "environment name": "zh-bd", "db backends": [], "email used for registration": "n/a"}, {"status": "no-server-assigned", "web backends": [], "server": "undefined", "environment name": "myproject-oldtest10", "db backends": [], "email used for registration": "peter.muster@gmail.com"}, {"status": "undefined", "web backends": [], "server": "undefined", "environment name": "acme2", "db backends": [], "email used for registration": "n/a"}, {"status": "undefined", "web backends": [], "server": "undefined", "environment name": "zh-vd", "db backends": [], "email used for registration": "n/a"}, {"status": "no-server-assigned", "web backends": [], "server": "undefined", "environment name": "myproject-oldtest1", "db backends": [], "email used for registration": "peter.muster@gmail.com"}, {"status": "no-server-assigned", "web backends": [], "server": "undefined", "environment name": "myproject-oldtest2", "db backends": [], "email used for registration": "peter.muster@gmail.com"}, {"status": "no-server-assigned", "web backends": [], "server": "undefined", "environment name": "myproject-oldtest3", "db backends": [], "email used for registration": "peter.muster@gmail.com"}, {"status": "activated", "web backends": ["pgrid0.myproject-old.com:10001"], "server": "pgrid0", "environment name": "myproject-oldtestjan2", "db backends": [], "email used for registration": "peter.muster@gmail.com"}, {"status": "activated", "web backends": ["pgrid0.myproject-old.com:10002"], "server": "pgrid0", "environment name": "09test1", "db backends": [], "email used for registration": "peter.muster@gmail.com"}, {"status": "activated", "web backends": ["pgrid0.myproject-old.com:10003"], "server": "pgrid0", "environment name": "testnew20151", "db backends": ["pgrid0.myproject-old.com:11003"], "email used for registration": "peter.muster@gmail.com"}], "servers": ["pgrid0"]}

On 01 Feb 2015, at 22:41, Varun Malhotra <notifications@github.com> wrote:

> In json2html/jsonconv.py:
> 
> >  		global a
> > -		a=a+ "<table border=\"1\">" 
> > +		global table_attributes
> > +
> > +		table_init_markup = "<table %s>" %(table_attributes)
> What about of no table_attributes is/are not present? Default table border has been overwritten. Also when I try the simple example:
> 
> _json2conv = {
>     "glossary": {
>         "title": "example glossary",
>         "GlossDiv": {
>             "title": "S",
>             "GlossList": {
>                 "GlossEntry": {
>                     "ID": "SGML",
>                     "SortAs": "SGML",
>                     "GlossTerm": "Standard Generalized Markup Language",
>                     "Acronym": "SGML",
>                     "Abbrev": "ISO 8879:1986",
>                     "GlossDef": {
>                         "para": "A meta-markup language, used to create markup languages such as DocBook.",
>                         "GlossSeeAlso": ["GML", "XML"]
>                     },
>                     "GlossSee": "markup"
>                 }
>             }
>         }
>     }
> }
> I didn't get the correct output. Please re-check what's being missing. Also, do share your json, I'm curious to see the sorting as suggested by you. It would be great to see everything working perfectly on my end. Thanks for introducing sort.
> 
> —
> Reply to this email directly or view it on GitHub.
>